cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringAgentService.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.tethering;
+
+import com.google.common.base.Preconditions;
+import com.google.gson.Gson;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.internal.remote.RemoteAuthenticator;
+import io.cdap.cdap.common.service.AbstractRetryableScheduledService;
+import io.cdap.cdap.common.service.RetryStrategies;
+import io.cdap.cdap.internal.app.store.AppMetadataStore;
+import io.cdap.cdap.spi.data.transaction.TransactionRunner;
+import io.cdap.cdap.spi.data.transaction.TransactionRunners;
+import io.cdap.common.http.HttpMethod;
+import io.cdap.common.http.HttpResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import javax.inject.Inject;
+
+/**
+ * The main class to run the remote agent service.
+ */
+public class TetheringAgentService extends AbstractRetryableScheduledService {
+  private static final Logger LOG = LoggerFactory.getLogger(TetheringAgentService.class);
+  private static final Gson GSON = new Gson();
+  private static final String CONNECT_CONTROL_CHANNEL = "/v3/tethering/controlchannels/";
+  private static final String SUBSCRIBER = "tether.agent";
+  private static final String TOPIC = "tethering";
+
+  private final CConfiguration cConf;
+  private final long connectionInterval;
+  private final TetheringStore store;
+  private final String instanceName;
+  private final TransactionRunner transactionRunner;
+  private String lastMessageId;
+
+  @Inject
+  TetheringAgentService(CConfiguration cConf, TransactionRunner transactionRunner) {
+    super(RetryStrategies.fixDelay(cConf.getLong(Constants.Tethering.CONNECT_INTERVAL), TimeUnit.SECONDS));
```

Comment:
Make the retry configurable via cConf. Please check with existing implementation on how to do so.

Reply:
If we use RetryStrageties.fromConfiguration(), then we'll need to specify max retries and max retry time configuration.  In our case, we probably want to retry forever, so it seems like these settings are not applicable. WDYT?

Reply:
fixed